### PR TITLE
Convert query_region to cone search

### LINKED ISSUE #585: Make SDSS query region a circle
This issue was discussed in #545, but we opted to do it as a separate fix so the other work in #545 could get in sooner.

Basically, the SDSS region queries (which have a `radius`) keyword in fact search a box instead of a circular region on the sky.  That needs to be fixed.  I think @weaverba137 _might_ be working on this?

### PR DESCRIPTION
**Pull request by weaverba137**
This PR closes #585.

`query_region()` now uses `query_crossid()` internally with slightly different parameters.  Effectively, a circular/cone region around each coordinate is searched instead of a rectangle in previous versions.

In addition, this PR adds lots of updates to tests, both online and offline, and cleans up the documentation of several of the methods.

Other comments:

* `query_region()` and `query_crossid()` accept `astropy.coordinates.Angle` for search radius as well as string or Quantity; internally, Angle is used to validate the search radius input and convert the angle to arcmin.
* The maximum search radius imposed by SDSS is 3 arcmin.  This is now enforced by the code.
* Added DR17 to offline tests; it had already been enabled for online tests.

### CODE REVIEW

**Comment by keflavich** [comment]
This is a pretty big PR that looks good at a glance.  @bsipocz, will you want to do a detailed review?  I'm ready to approve this, but I haven't checked every line of the tests.

**Comment by bsipocz** [comment]
Yes, I definitely want to review this with an ETA of maybe next week as this one is already been rather full.

File: astroquery/sdss/core.py
Comment (eerovaher) [inline astroquery/sdss/core.py:147]: `astropy` has a fitting function for this: https://github.com/astropy/astropy/blob/c33e53f68388d4b4d68e85956bcfe4500a94c642/astropy/utils/misc.py#L45 The only case that needs to be handled separately is `str` because every `str` is iterable.

File: astroquery/sdss/core.py
Comment (eerovaher) [inline astroquery/sdss/core.py:150]: Why not use f-strings here?
```suggestion
            obj_names = [f'obj_{i}' for i in range(len(coordinates))]
```

File: astroquery/sdss/core.py
Comment (eerovaher) [inline astroquery/sdss/core.py:153]: I like very explicit error messages.
```suggestion
            raise ValueError(
                f"number of coordinates {len(coordinates)} is not equal to number of object names {len(obj_names)}"
            )
```

File: astroquery/sdss/core.py
Comment (eerovaher) [inline astroquery/sdss/core.py:163]: The `list` created here is discarded after iterating over it once, so it would be better to use a generator expression here. Furthermore, using `zip` is preferable over accessing elements by index.
```suggestion
        data += " \n ".join(data_format.format(obj=obj,
                                               ra=coord.ra.deg,
                                               dec=coord.dec.deg)
                             for obj, coord in zip(obj_names, coordinates))
```

File: astroquery/sdss/core.py
Comment (eerovaher) [inline astroquery/sdss/core.py:1079]:
```suggestion
        sql = f"{q_select} {q_from} {q_where}"
```

File: astroquery/sdss/tests/test_sdss.py
Comment (eerovaher) [inline astroquery/sdss/tests/test_sdss.py:315]: I suggest letting `pytest` [check the error message with the `match` argument](https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions) to ensure you are indeed getting the `ValueError` you are expecting. Furthermore, the error message often conveys the purpose of the test clearly enough that an additional comment is not needed.

**Review by eerovaher — COMMENTED** [review]
I have a few comments, but I wont claim to have done a thorough review.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:83]: True, but not every `Quantity` is an acceptable input either.  One way or another, it has to be an angular quantity.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:61]: OK, I can rearrange.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:117]: I agree, but for historical reasons, that is how this is implemented. Perhaps changing the API in that manner should be a separate PR.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:133]: If the radius is already an Angle, there is no need to test whether it is convertible to Angle.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:150]: Historical reasons.  Is there an active policy in astroquery to convert to f-strings wherever possible?

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:163]: I was trying to make a minimal change.  Again, is there an explicit policy that replacing explicit lists with implicit iterables (which are less obvious) should be undertaken wherever possible?

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:153]: I was trying to make a minimal change. My personal experience is that PR reviewers are more likely to say "don't change what isn't broken".

File: astroquery/sdss/tests/test_sdss.py
Comment (weaverba137) [reply astroquery/sdss/tests/test_sdss.py:315]: I'll take a look at `match`, but honestly, I'm going to leave the docstring.

File: astroquery/sdss/core.py
Comment (bsipocz) [reply astroquery/sdss/core.py:61]: please do not rearrange, rather make them mandatory kwarg. We incrementally doing this change in other modules, and examples like these are excellent cases where optional parameters should be forced to spelled out. 
(Of course it will be an API break, but I plan to have multiple API breaks in the next release anyway by doing significant cleanups).

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:147]: I'll give that a try.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:137]: Is it measurably faster on a scalar?

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:1079]: Again, minimal change.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:61]: Is there a good example of that for me to look at?

Comment (bsipocz) [reply astroquery/sdss/core.py:61]: Yeap, here is the PEP: https://peps.python.org/pep-3102/

For astroquery e.g. alma, esasky, mast are all moved on this approach.

In practice you add a `*` between positional and keyword arguments, and then can rearrange the order of the keyword ones as you wish (I fully agree that in many cases it makes sense to rearrange them, but if we don't enforce them to be keyword only, any addition is a breaking API change.)

File: astroquery/sdss/core.py
Comment (bsipocz) [reply astroquery/sdss/core.py:150]: we incrementally cleaning up old python syntax, and as f-strings are cleaner to read moving towards using them more.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:61]: I think using keyword-only arguments is a good idea, but it might be better to apply that change to the entire `sdss` subpackage at once instead of doing it one function at a time.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:117]: If this is how it works already then a separate pull request would indeed be preferable.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:133]: Yes, but the previous code also worked with a bare `Quantity` without trying to construct an `Angle` from it.

Comment (weaverba137) [reply astroquery/sdss/core.py:133]: And I have not eliminated that functionality, if you examine the entire function.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:137]: Using `IPython`:
```python
>>> from astropy import units as u
>>> angle = 1 * u.deg
>>> %timeit angle.to(u.rad).value
8.11 µs ± 199 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
>>> %timeit angle.to_value(u.rad)
4.58 µs ± 170 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:150]: Will do.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:153]: If this is an already existing error message then it should be modified in a separate pull request.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:147]: In the interest of getting this PR merged, I'm *not* going to attempt to change this, in the interest of minimizing additional changes.

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:163]: Pull requests are not required to update unrelated code, but these lines are being touched anyways.

Replacing list comprehensions with generator expressions is faster and uses less memory, but the values can be accessed only once. In this case the values are accessed only once anyways, so there really are no downsides to using a generator expression here.

Combining iterables with `zip` is faster than looping over indices. Using `IPython`:
```python
>>> word = "foo"
>>> letters = list(word)
>>> %timeit for i in range(len(word)): assert word[i] == letters[i]
392 ns ± 32.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
>>> %timeit for x, y in zip(word, letters): assert x == y
280 ns ± 2.81 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

By using an f-string the code snippet could be made even shorter:
```suggestion
        data += " \n ".join(f"{obj} {coord.ra.deg} {coord.dec.deg}" for obj, coord in zip(obj_names, coordinates))
```

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:1079]: These lines are being edited anyways and the change is not user-facing.

Comment (weaverba137) [reply astroquery/sdss/core.py:1079]: Done

File: astroquery/sdss/core.py
Comment (eerovaher) [reply astroquery/sdss/core.py:133]: The code still works with a bare `Quantity`, but it now converts a bare `Quantity` to an `Angle` even though none of the additional functionality implemented by `Angle` is required.

**Comment by weaverba137** [comment]
In regards keyword-only arguments: I have converted the entire module except for the `get_images` and `get_spectra` functions, which have a fairly complicated call signature.  I think that is sufficient for the current PR, given that the original target of the PR was to convert to cone-search, not specifically to change to keyword-only arguments.

**Review by bsipocz — APPROVED** [review]
Overall it look good, bar one comment about the removal of `get_query_payload`, which I would like to be back, and to e.g. return the SQL string.
(It proved to be rather useful in many cases in the past.)

Anyway, I don't think it's worth dragging this largeish PR with that point further and will open an issue for a follow-up instead.

As the other review pointed out, there are a few places where a little bit of cleanup could be done, but I agree that that one is definitely beyond the scope here.

**Comment by bsipocz** [comment]
Thanks @weaverba137!

**Comment by bsipocz** [comment]
After merging this PR, I've gone through and triaged the `sdss` labelled issues, and happy to say it has fixed a bunch of them. There are still some long-reported bugs, but after the triage, we have half the number of open issues than before! 

🎉

**Comment by pllim** [comment]
Thanks!

**Comment by emirkmo** [comment]
Just out of curiosity, is it still possible to use astroquery to make larger radius queries similar to the box version previously used by astroquery.query_region?

You know how the saying goes with old code and user workflows…, but we had built an automated pipeline around large box size crossmatches 😅. 

the limit of 3 arcmin is way too small and needs a hefty workaround based on SDSS fields..

we can currently pin the old astroquery version and patch in future fixes that don’t touch this code but it’s not ideal.

 As a note from a user, it would have been nice to get a depreciation warning for an API breaking change like this, one that lasted a little while to give us time to adjust. It seems to me the switchover happened rapidly, and we have to do workarounds such as building the old functionality from the git history ourselves and patching in future fixes, or switch to our own SQL queries/casjobs. 
 
One way to enable backwards compatibility is to rename the old version to something like query_rectangle_region or to add a parameter such as rectangular: boolean = False, that preserves the breakage while the deprecation warning is active.

Anyway, hope this can be considered.

**Comment by bsipocz** [comment]
@emirkmo - you can write any custom SQL query, so I would suggest looking into using `query_sql()` for larger sized queries?

**Comment by emirkmo** [comment]
Thanks for the advice. It is actually easy to solve as you say, but moving to a custom SQL query is exactly what I was hoping to avoid by using astroquery. I made an issue to track what I guess is actually a feature request to re-enable to rectangular search, as it has a tangible benefit, and the old implementation in astroquery from a few months ago still just works.

**Comment by pllim** [comment]
I am also interested in this request because the 3 arcmin limit seems pretty arbitrary and I am not sure how to explain to users why suddenly we have this limit downstream.

**Comment by bsipocz** [comment]
cc @weaverba137

**Comment by bsipocz** [comment]
Frankly, we can still roll back that API change (it hasn't landed in a tagged release), but I would need a bit more feedback (doing the proper conesearch was raised as a request a few times in the past...)


cc @keflavich

**Comment by weaverba137** [comment]
3 arcmin is a *default*, but it is not *hard-coded* as far as I'm aware. What happens when you increase that limit?

**Comment by keflavich** [comment]
I've been listening in on this, and I'd like to hear @weaverba137 's take.  I think rolling back the API change makes sense; probably best would be to allow either rectangular or circular searches, which is what we do with some other modules (vizier, I think?  I'd have to dig more).

Generally I agree that, for something as basic as a rectangle or circle query, we want to provide the simplified user interface and avoid having users write SQL.

**Comment by bsipocz** [comment]
OK, but here the underlying difference is huge. E.g. for the rectangular one we just wrote a brute force +/- ra/dec box, while the radius search is using a different server-side service.

**Comment by emirkmo** [comment]
> Frankly, we can still roll back that API change (it hasn't landed in a tagged release), but I would need a bit more feedback (doing the proper conesearch was raised as a request a few times in the past...)
> 
> cc @keflavich

I think the changes are actually very nice, especially with more proper reporting of errors and test failures. Run query also works as expected now. I would just appreciate having access to the rectangular search as well. I will leave any decisions up to the maintainers of course.

**Comment by bsipocz** [comment]
I would see value in not having API change for the old setup, but then we run into the core of the previous issues, e.g. that `query_region` is not actually a cone search, but a box while a cone search is available. Adding a kwarg to fall back to the previous behaviour is still an API change, so while semantically would work, I'm looking for other options, too, maybe the we need a new function name for the current cone search-based implementation. (but I don't have any good ideas how to call it.)

**Comment by emirkmo** [comment]
`query_cone_region`

`query_rectangle_region`

Is what I would look for in the docs (and first looked for).

Where either one or the other just calls `query_region` under the hood. That depends on whether backwards compatibility is more important or whether the function arguments making sense is more important.

The docstring for either can point to each other and mention the different endpoints used and the radius limitation.

**Comment by bsipocz** [comment]
Yeah, let's wait for Ben's take and then we'll figure out something that brings in the least amount of API change.

**Comment by keflavich** [comment]
If we do stick with the API change, we should include a deprecation warning that points users to the right way to do things.

**Comment by bsipocz** [comment]
I'm starting to incline towards adding a kwarg to `query_region` that defaults to rectangular. Quite ugly, but that would be seamless, and no API change for all the previous use cases (I totally missed this server-side limitation on the query size, so the change doesn't feel like a simple backend refactoring any more to make the code more efficient and to fix a bug)

**Comment by weaverba137** [comment]
Before I reply fully, I'm not entirely sure this is the thread we should continue the discussion on. For one, it's harder to find in the GitHub system since it's closed. Another open issue or PR might be easier for everyone to see.

**Comment by emirkmo** [comment]
> Before I reply fully, I'm not entirely sure this is the thread we should continue the discussion on. For one, it's harder to find in the GitHub system since it's closed. Another open issue or PR might be easier for everyone to see.

I opened #2659

**Comment by weaverba137** [comment]
Unless anyone objects, I am happy to move over to #2659.

**Comment by bsipocz** [comment]
Yes, please move this convo to the issue linked above!

**Pull request merged**
